### PR TITLE
REST API: Themes: Expose some additional basic fields

### PR DESCRIPTION
**Pull request by ockham**
The `/wp/v2/themes` [endpoint](https://developer.wordpress.org/rest-api/reference/themes/) currently only exposes each theme's theme_supports field, but nothing else -- not even theme name or title.

See https://core.trac.wordpress.org/ticket/49906.

Needed for https://github.com/WordPress/gutenberg/pull/21578 -- you can use that PR to test this one :slightly_smiling_face: I'm adding all the fields that I think will be needed to implement https://github.com/WordPress/gutenberg/issues/20469#issuecomment-597806801:

image

That is:
- `author`
- `author_name`
- `author_uri`
- `description`
- `name`
- `screenshot`
- `stylesheet`
- `template`
- `theme_uri`
- `version`

Implementation loosely inspired by https://github.com/Automattic/jetpack/blob/f6f313d268c336491c8bca1afcf3e5d3663597ce/json-endpoints/jetpack/class.jetpack-json-api-themes-endpoint.php.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by TimothyBJacobs** [comment]
Looks great! Are you interested in adding tests for this?

**Comment by ockham** [comment]
> Looks great! Are you interested in adding tests for this?

Thanks! Sure, I'll add some tomorrow :smile:

**Comment by ockham** [comment]
Update, I've fixed the failing unit tests and lint.

I haven't yet added new tests, since I'm still struggling with setting up the tests environment on my Linux box. I asked on Core Slack: https://wordpress.slack.com/archives/C02RQBWTW/p1586985925240400

**Comment by ockham** [comment]
I managed to get unit tests to run locally, and added a few more fields and corresponding tests (see updated PR description). I'd say this is ready for review :slightly_smiling_face:

**Comment by ockham** [comment]
cc/ @mcsf for review. Because it's a themes endpoint! :tada:

**Comment by mcsf** [comment]
> Because it's a themes endpoint! 🎉

Good times!

Looks great, as far as I'm concerned. I'd defer to Timothy or K Adam for the final green light, though.

**Comment by ockham** [comment]
> Looks great, as far as I'm concerned. I'd defer to Timothy or K Adam for the final green light, though.

@kadamwhite @TimothyBJacobs Pretty pleaaase? :blush:

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php [hunk 1]
```diff
@@ -115,6 +115,28 @@ public function prepare_item_for_response( $theme, $request ) {
...
+		);
+		$fields_to_include = array_intersect( array_keys( $field_mappings ), $fields );
+
+		foreach ( $fields_to_include as $field ) {
+			$data[ $field ] = $theme[ $field_mappings[ $field ] ];
+		}
+
+		if ( in_array( 'screenshot', $fields, true ) ) {
+			// Using $theme->get_screenshot() with no args to get absolute URL.
+			$data['screenshot'] = $theme->get_screenshot();
```
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:137]: Let's make sure this returns an empty string, not false, when the theme does not have a screenshot.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php [hunk 2]
```diff
@@ -115,6 +115,28 @@ public function prepare_item_for_response( $theme, $request ) {
...
+			'name'        => 'Name',
+			'stylesheet'  => 'Stylesheet',
+			'template'    => 'Template',
+			'theme_uri'   => 'Theme URI',
+			'version'     => 'Version',
+		);
+		$fields_to_include = array_intersect( array_keys( $field_mappings ), $fields );
+
+		foreach ( $fields_to_include as $field ) {
+			$data[ $field ] = $theme[ $field_mappings[ $field ] ];
```
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:132]: This ends up giving us a mixture of translated/untranslated and markedup/unmarked up properties.

I'd prefer it if we followed the `raw/rendered` format of the other REST API endpoints. So that would look something more like this.

```php
$data['author'] = array(
    'raw' => $theme->display('Author', false, true ),
    'rendered' => $theme->display('Author' ),
);
```

We'd follow that pattern for `Author`, `Author URI`, `Description`, `Name`, and `Theme URI`. And we would drop the separate `Author Name` field entirely.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php [hunk 3]
```diff
@@ -192,6 +214,46 @@ public function get_item_schema() {
...
+				),
+				'author_name'    => array(
+					'description' => __( 'The theme author\'s name.' ),
+					'type'        => 'string',
+					'readonly'    => true,
+				),
+				'author_uri'     => array(
+					'description' => __( 'The website of the theme author.' ),
+					'type'        => 'string',
+					'readonly'    => true,
```
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:230]: Let's add `format: 'uri'` here.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php [hunk 4]
```diff
@@ -458,6 +520,16 @@ public function get_item_schema() {
 						),
 					),
 				),
+				'theme_uri'      => array(
+					'description' => __( 'The URI of the theme\'s webpage.' ),
+					'type'        => 'string',
+					'readonly'    => true,
```
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:526]: And `format: 'uri'` here as well.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php (hunk 3 above)
Comment (ockham) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:231]: 9282f21fe74989b06b5b3d359d281ea595c082e7

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php (hunk 4 above)
Comment (ockham) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:527]: 9282f21fe74989b06b5b3d359d281ea595c082e7

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php (hunk 1 above)
Comment (ockham) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php]: 9b9bace07e3798930406aa48d29be99984f86999

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php (hunk 2 above)
Comment (ockham) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:132]: Good point, thanks!

This works great for `author`; e.g. for the default theme:

```js
{
	raw: 'Michael Heilemann',
	rendered: '<a href="http://binarybonsai.com/">Michael Heilemann</a>'
}
```

I'm not so sure about some of the others. E.g. do we really want a `rendered` `AuthorURI`? Should that be `<a href="http://binarybonsai.com/">http://binarybonsai.com/</a>`? IMO, the `rendered` `Author` makes more sense as link text. This raises the question of whether to keep `AutorURI` around as a 'plain' string field, or whether to drop it. (I'm keeping it for now, as it seems useful in a client to build a clickable link from `author_uri` and `raw` `author` without needing to sanitize `rendered` `author`.)

The same argument largely applies to `ThemeURI`, where a clickable link could be built from that field (if left as a plain string) and `Name`.

Furthermore, I'm not sure if `$theme->display()` is working so well for a number of fields; `Description` always seems to contain a link for me (again for the default theme):

```
4) WP_Test_REST_Themes_Controller::test_theme_description
Failed asserting that Array &0 (
    'raw' => 'The default WordPress theme based on the famous <a href="http://binarybonsai.com/kubrick/">Kubrick</a>.'
    'rendered' => 'The default WordPress theme based on the famous <a href="http://binarybonsai.com/kubrick/">Kubrick</a>.'
) is identical to [...]
```

Finally, what about `Name`? Is markup allowed in that field? If no, I'd also stick with a plain string :thinking:

Comment (TimothyBJacobs) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:132]: Using `display` means we'd be using the existing [`WP_Theme::markup_header`](https://github.com/WordPress/wordpress-develop/blob/86cbde382a0d323480a14b53c86f274599880ee2/src/wp-includes/class-wp-theme.php#L885). We wouldn't be making those urls links, WP applies `esc_url` to those values.

For the description, the `<a>` tag is there because it is present in the [theme's actual description](https://themes.trac.wordpress.org/browser/default/1.7.2/style.css#L4). It isn't be added as an artifact of using `display`. 

> Finally, what about Name? Is markup allowed in that field? If no, I'd also stick with a plain string 🤔

Name get's a fallback set when using `display`.

Comment (ockham) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:130]: Ah, gotcha. Okay, I'll use `display` for those :+1:

Comment (ockham) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php]: 8b16987

**Comment by ockham** [comment]
Addressed all feedback. Should be ready for another look!

**Comment by TimothyBJacobs** [comment]
Thanks, looking good!

One small thing, we should figure out some better language for `as it exists in the database.` These values aren't stored in the database, they are pulled from the theme's files.

The values are also sanitized and translated, so I'm not sure we could say `as it exists in the theme's style.css` file either. Not sure  ¯\_(ツ)_/¯.

Also tagging @spacedmonkey for a general review as he's expressed interest in expanding the themes endpoint.

**Comment by ockham** [comment]
> Thanks, looking good!
> 
> One small thing, we should figure out some better language for `as it exists in the database.` These values aren't stored in the database, they are pulled from the theme's files.

Ah yeah, good point. Lazy copy-pasta from another endpoint :grimacing: 
 
> The values are also sanitized and translated, so I'm not sure we could say `as it exists in the theme's style.css` file either. Not sure ¯_(ツ)_/¯.

I'll think of something :+1: 

> Also tagging @spacedmonkey for a general review as he's expressed interest in expanding the themes endpoint.

:+1:

**Comment by ockham** [comment]
I changed the wording to `as found in the theme header`. That still doesn't mention sanitization, translation, or even fallbacks (in case of `Name`), but maybe it's good enough?

**Comment by TimothyBJacobs** [comment]
That wording sounds good to me!

**Comment by ockham** [comment]
:wave: @TimothyBJacobs @spacedmonkey 

Just checking in if there's anything else I should do or if mayyybe I could get y'all's approval :blush:

**Comment by TimothyBJacobs** [comment]
This still looks good to me @ockham! Will commit later today/tomorrow unless I hear different from @spacedmonkey.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php [hunk 5]
```diff
@@ -115,6 +115,40 @@ public function prepare_item_for_response( $theme, $request ) {
...
+			$data['screenshot'] = $theme->get_screenshot() ?: '';
+		}
+
+		$rich_field_mappings = array(
+			'author'      => 'Author',
+			'author_uri'  => 'AuthorURI',
+			'description' => 'Description',
+			'name'        => 'Name',
+			'theme_uri'   => 'ThemeURI',
+		);
```
Comment (spacedmonkey) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:141]: Missing fields  `'Tags'  ,'TextDomain' ,DomainPath' ,'RequiresWP' ,'RequiresPHP'`

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php [hunk 6]
```diff
@@ -192,6 +226,90 @@ public function get_item_schema() {
...
+						'rendered' => array(
+							'description' => __( 'HTML for the theme author, transformed for display.' ),
+							'type'        => 'string',
+							'readonly'    => true,
+						),
+					),
+				),
+				'author_uri'     => array(
+					'description' => __( 'The website of the theme author.' ),
+					'format'      => 'uri',
```
Comment (spacedmonkey) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:248]: Is this the correct place for a uri?

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php [hunk 7]
```diff
@@ -192,6 +226,90 @@ public function get_item_schema() {
...
+					'type'        => 'string',
+					'readonly'    => true,
+					'properties'  => array(
+						'raw'      => array(
+							'description' => __( 'The website of the theme author, as found in the theme header.' ),
+							'type'        => 'string',
```
Comment (spacedmonkey) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:254]: should the format of uri be here?

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php [hunk 8]
```diff
@@ -192,6 +226,90 @@ public function get_item_schema() {
...
+							'description' => __( 'HTML for the theme author, transformed for display.' ),
+							'type'        => 'string',
+							'readonly'    => true,
+						),
+					),
+				),
+				'author_uri'     => array(
+					'description' => __( 'The website of the theme author.' ),
+					'format'      => 'uri',
+					'type'        => 'string',
```
Comment (spacedmonkey) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:249]: This needs to be an object here as properties are defined.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php [hunk 9]
```diff
@@ -192,6 +226,90 @@ public function get_item_schema() {
...
+						'rendered' => array(
+							'description' => __( 'The website of the theme author, transformed for display.' ),
+							'type'        => 'string',
+							'readonly'    => true,
+						),
+					),
+				),
+				'description'    => array(
+					'description' => __( 'A description of the theme.' ),
+					'type'        => 'string',
```
Comment (spacedmonkey) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:266]:
```suggestion
					'type'        => 'object',
```

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php [hunk 10]
```diff
@@ -192,6 +226,90 @@ public function get_item_schema() {
...
+						'rendered' => array(
+							'description' => __( 'The theme description, transformed for display.' ),
+							'type'        => 'string',
+							'readonly'    => true,
+						),
+					),
+				),
+				'name'           => array(
+					'description' => __( 'The name of the theme.' ),
+					'type'        => 'string',
```
Comment (spacedmonkey) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:283]:
```suggestion
					'type'        => 'object',
```

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php [hunk 11]
```diff
@@ -458,6 +576,29 @@ public function get_item_schema() {
 						),
 					),
 				),
+				'theme_uri'      => array(
+					'description' => __( 'The URI of the theme\'s webpage.' ),
+					'type'        => 'string',
```
Comment (spacedmonkey) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:581]:
```suggestion
					'type'        => 'object',
```

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php [hunk 12]
```diff
@@ -192,6 +226,90 @@ public function get_item_schema() {
...
+					'readonly'    => true,
+					'properties'  => array(
+						'raw'      => array(
+							'description' => __( 'The website of the theme author, as found in the theme header.' ),
+							'type'        => 'string',
+							'readonly'    => true,
+						),
+						'rendered' => array(
+							'description' => __( 'The website of the theme author, transformed for display.' ),
+							'type'        => 'string',
```
Comment (spacedmonkey) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:259]: uri should be here

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php [hunk 13]
```diff
@@ -192,6 +226,90 @@ public function get_item_schema() {
 			'title'      => 'theme',
 			'type'       => 'object',
 			'properties' => array(
+				'author'         => array(
+					'description' => __( 'The theme author.' ),
+					'type'        => 'string',
```
Comment (spacedmonkey) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:231]:
```suggestion
					'type'        => 'object',
```

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php (hunk 8 above)
Comment (spacedmonkey) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:249]:
```suggestion
					'type'        => 'object',
```

Comment (TimothyBJacobs) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:249]: Yeah, great catch here. It's a `raw`, `render` like the other properties.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php (hunk 6 above)
Comment (TimothyBJacobs) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:248]: Yeah `uri` is the right format unless I'm missing something.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php (hunk 7 above)
Comment (TimothyBJacobs) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:254]: Yeah `uri` is the right format unless I'm missing something.

**Comment by spacedmonkey** [comment]
At the moment, these fields can only display data on the currently active theme.

With a very small tweak, to get_items, we could display this data for all themes. 
```
	public function get_items( $request ) {
		// Retrieve the list of registered collection query parameters.
		$registered = $this->get_collection_params();
		$themes     = array();

		if ( isset( $registered['status'], $request['status'] ) && in_array( 'active', $request['status'], true ) ) {
			$active_theme = wp_get_theme();
			$active_theme = $this->prepare_item_for_response( $active_theme, $request );
			$themes[]     = $this->prepare_response_for_collection( $active_theme );
		} else {
                         $active_themes = wp_get_themes();
                         foreach( $active_themes as $theme ) {
			     $active_theme = $this->prepare_item_for_response( $theme, $request );
			     $themes[]     = $this->prepare_response_for_collection( $active_theme );
                         }
                }

		$response = rest_ensure_response( $themes );

		$response->header( 'X-WP-Total', count( $themes ) );
		$response->header( 'X-WP-TotalPages', count( $themes ) );

		return $response;
	}
```

The only issue with this, is that theme supports can not be displayed on themes that are not active.

**Comment by TimothyBJacobs** [comment]
> At the moment, these fields can only display data on the currently active theme.

> With a very small tweak, to get_items, we could display this data for all themes.

I think it makes sense to expand that in 5.5, but let's make it a separate ticket.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php (hunk 5 above)
Comment (TimothyBJacobs) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:141]: +1 to including these. `tags` should follow the `raw`, `rendered` format.

**Comment by spacedmonkey** [comment]
I have created another PR for showing all themes in the API - https://github.com/WordPress/wordpress-develop/pull/264

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php (hunk 5 above)
Comment (spacedmonkey) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:141]: There are also Parent Theme and template fields we should think about as well.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php (hunk 8 above)
Comment (ockham) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:249]: Great catch indeed! :sweat_smile:

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php (hunk 6 above)
Comment (ockham) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:248]: See https://github.com/WordPress/wordpress-develop/pull/222#discussion_r423992124 :sweat_smile:

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php (hunk 9 above)
Comment (ockham) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:266]: Change `type` to `object` for `description`

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php (hunk 11 above)
Comment (ockham) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:581]: Change `type` to `object` for `theme_uri`

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php (hunk 7 above)
Comment (ockham) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php]: 769a614

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php (hunk 12 above)
Comment (ockham) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php]: 6c0bcfd4b7

Comment (TimothyBJacobs) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php]: To clarify, the `type` should still be `string` but have a `format => uri`.

**Comment by TimothyBJacobs** [comment]
Hey @ockham just checking in on this!

**Comment by ockham** [comment]
> Hey @ockham just checking in on this!

Thanks! I'm back today from 2 weeks of maintenance rotation, picking this up again :+1:

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php (hunk 5 above)
Comment (ockham) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:142]: > Missing fields `'Tags' ,'TextDomain' ,DomainPath' ,'RequiresWP' ,'RequiresPHP'`

I just added `Tags` (34a179c). I've noticed that the other ones are _header_ fields (i.e. taken from the theme's `style.css` header); however, they're optional and quite 'technical' -- it's conceivable that a REST API client might not be that interested in the required WordPress and PHP versions (if it's e.g. not even programmed in PHP). A similar point can be made for text domain/i18n related fields.

And then, as you mentioned [in your other comment](https://github.com/WordPress/wordpress-develop/pull/222#discussion_r424615493), there's also [these fields](https://github.com/WordPress/wordpress-develop/blob/e18d6e942b7bbc2b776d922d7c9657c4ab1c58b4/src/wp-includes/class-wp-theme.php#L558-L578) (the 'offset' ones, i.e. the ones that can be accessed through square-bracket/pseudo-associative-array syntax), which arguably reflect the public API of the `WP_Theme` class better).

My question is, do we want to reflect
1. the header fields
2. the 'offset' fields
3. the union of 1. and 2.?

I'd personally lean towards option 2.

If this needs more discussion: Would we consider merging this PR pretty much as-is to provide some basic fields, and add the other ones in a follow-up once we've reached consensus?

Comment (TimothyBJacobs) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:142]: Tags looks good! Let's also add an `items` declaration that specifies each item is a `string` type.

> they're optional and quite 'technical' -- it's conceivable that a REST API client might not be that interested in the required WordPress and PHP versions (if it's e.g. not even programmed in PHP). A similar point can be made for text domain/i18n related fields.

I don't disagree that there are some clients that might not be interested in that data, but there are plenty of fields some clients might not find useful. If they want to restrict the fields returned the fields they need, they can always use the `_fields` parameter. This information is helpful when querying the list of themes in a site. And will be absolutely necessary when we allow switching themes via the REST API.

We are including the text domain in the block types endpoint. Let's leave out domain path for now, since there are alternate ways we could display that and it can't really be utilized.

> And then, as you mentioned in your other comment, there's also these fields 

Most of these we have implemented. The remaining are the parent theme and files. Let's exclude the template/css files for now since those aren't currently public information and will need more auth consideration. For the parent theme, let's omit that until we determine how exactly returning all themes will work.

> which arguably reflect the public API of the WP_Theme class better.

I'm not sure I'd agree with this. I think the named method calls are the preferred way to interact with that data in a meaningful manner. The offsets exist primarily for back-compat with the previous data structure. [Trac Ticket](https://core.trac.wordpress.org/ticket/20103#comment:9).

> If this needs more discussion: Would we consider merging this PR pretty much as-is to provide some basic fields, and add the other ones in a follow-up once we've reached consensus?

My goal will be to commit this this week, if that means we need to punt on some things, we can punt on them, but I think we can probably figure out the changes we need before then.

Comment (ockham) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:154]: > Tags looks good! Let's also add an `items` declaration that specifies each item is a `string` type.

Thanks, added per 34a179c. BTW I tried searching the REST API docs for an overview of schema props and types but couldn't really find a comprehensive guide. Would you have a pointer for me?

> > they're optional and quite 'technical' -- it's conceivable that a REST API client might not be that interested in the required WordPress and PHP versions (if it's e.g. not even programmed in PHP). A similar point can be made for text domain/i18n related fields.
> 
> I don't disagree that there are some clients that might not be interested in that data, but there are plenty of fields some clients might not find useful. If they want to restrict the fields returned the fields they need, they can always use the `_fields` parameter. This information is helpful when querying the list of themes in a site. And will be absolutely necessary when we allow switching themes via the REST API.

Good point :+1: 

> We are including the text domain in the block types endpoint. Let's leave out domain path for now, since there are alternate ways we could display that and it can't really be utilized.

Thanks! Including `texdomain` per 4a959aee182893a35fbadb78d5330e3e1e48b12d.

> > And then, as you mentioned in your other comment, there's also these fields
> 
> Most of these we have implemented. The remaining are the parent theme and files. Let's exclude the template/css files for now since those aren't currently public information and will need more auth consideration. For the parent theme, let's omit that until we determine how exactly returning all themes will work.

:+1: 

> > which arguably reflect the public API of the WP_Theme class better.
> 
> I'm not sure I'd agree with this. I think the named method calls are the preferred way to interact with that data in a meaningful manner. The offsets exist primarily for back-compat with the previous data structure. [Trac Ticket](https://core.trac.wordpress.org/ticket/20103#comment:9).

Ah, thanks! This has been enlightening. I've taken the opportunity to switch the controller to use `$theme->get()` (which is basically a wrapper around the named methods) rather than array access, d14da4093bc35259e6b74640647378ac8ae1950e.

> > If this needs more discussion: Would we consider merging this PR pretty much as-is to provide some basic fields, and add the other ones in a follow-up once we've reached consensus?
> 
> My goal will be to commit this this week, if that means we need to punt on some things, we can punt on them, but I think we can probably figure out the changes we need before then.

Sounds great!

To recap, I'm reading your comment as agreeing to hold off on adding most fields for now, except for `textdomain` (which I've now added). Please let me know if I was misreading you and if you'd like to see any other fields included!

Comment (TimothyBJacobs) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:154]: > BTW I tried searching the REST API docs for an overview of schema props and types but couldn't really find a comprehensive guide. Would you have a pointer for me?

We follow JSON Schema version 4, so that'd be the term to search for. The [spec]( https://json-schema.org/specification-links.html#draft-4) is a decent resource. I also wrote an [article](https://timothybjacobs.com/2017/05/17/json-schema-and-the-wp-rest-api/) that I think is decent, ignore the bits about using a separate schema validator.

> Thanks! Including `texdomain` per `4a959ae`.

Looks good!

> I've taken the opportunity to switch the controller to use `$theme->get()`

Looks good!

> Please let me know if I was misreading you and if you'd like to see any other fields included!

I think we should still include the requires PHP and WP headers.

Comment (ockham) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:156]: > > BTW I tried searching the REST API docs for an overview of schema props and types but couldn't really find a comprehensive guide. Would you have a pointer for me?
> 
> We follow JSON Schema version 4, so that'd be the term to search for. The [spec](https://json-schema.org/specification-links.html#draft-4) is a decent resource. I also wrote an [article](https://timothybjacobs.com/2017/05/17/json-schema-and-the-wp-rest-api/) that I think is decent, ignore the bits about using a separate schema validator.

Thanks for those pointers!

> > Please let me know if I was misreading you and if you'd like to see any other fields included!
> 
> I think we should still include the requires PHP and WP headers.

Done in fac13e5 :+1:

Comment (TimothyBJacobs) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:156]: lgtm!

Comment (ockham) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:156]: I _think_ this was the last feedback thread to address. So maybe :crossed_fingers: we're on track for approval? :smile:

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php (hunk 12 above)
Comment (TimothyBJacobs) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php]: @ockham this bit.

**Comment by TimothyBJacobs** [comment]
@ockham To clarify, the type should still be string but have a format => uri.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php (hunk 12 above)
Comment (ockham) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php]: Oops, missed this one. On it!

Comment (ockham) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php:275]: 7f5db79e2b

**Comment by ockham** [comment]
> @ockham To clarify, the type should still be string but have a format => uri.

Fixed by 7f5db79 :slightly_smiling_face:

**Comment by TimothyBJacobs** [comment]
Merged in [r47921](https://core.trac.wordpress.org/changeset/47921).
